src/main/java/ru/yandex/practicum/filmorate/storage/review/impl/ReviewDbStorage.java

```diff
@@ -0,0 +1,124 @@
+package ru.yandex.practicum.filmorate.storage.review.impl;
+
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.support.GeneratedKeyHolder;
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Component;
+import org.springframework.stereotype.Repository;
+import ru.yandex.practicum.filmorate.model.Review;
+import ru.yandex.practicum.filmorate.storage.review.ReviewStorage;
+
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.List;
+
+@Repository
+@Component("ReviewDbStorage")
+public class ReviewDbStorage implements ReviewStorage {
+    private final JdbcTemplate jdbcTemplate;
+
+    public ReviewDbStorage(JdbcTemplate jdbcTemplate) {
+        this.jdbcTemplate = jdbcTemplate;
+    }
+
+    public long addReview(Review review) {
+        String sqlQuery = "INSERT INTO reviews (user_id, film_id, content, is_positive) VALUES (?, ?, ?, ?)";
+        KeyHolder keyHolder = new GeneratedKeyHolder();
+
+        jdbcTemplate.update(connection -> {
+            PreparedStatement ps = connection.prepareStatement(sqlQuery, new String[]{"review_id"});
+            ps.setLong(1, review.getUserId());
+            ps.setLong(2, review.getFilmId());
+            ps.setString(3, review.getContent());
+            ps.setBoolean(4, review.getIsPositive());
+            return ps;
+        }, keyHolder);
+
+        return keyHolder.getKey().longValue();
+
+    }
+
+    public Review getReviewById(Long id) {
+        String sqlQuery = "SELECT r.review_id, r.film_id, r.user_id, r.content, r.is_positive," +
+                "COUNT(postive_r.review_id) - COUNT(negative_r.review_id) AS useful " +
+                "FROM reviews AS r " +
+                "LEFT JOIN (SELECT review_id FROM reviews_reactions WHERE is_positive = TRUE) AS postive_r ON r.review_id = postive_r.review_id " +
+                "LEFT JOIN (SELECT review_id FROM reviews_reactions WHERE is_positive = FALSE) AS negative_r ON r.review_id = negative_r.review_id " +
+                "WHERE r.review_id = ? " +
+                "GROUP BY r.review_id, r.film_id, r.user_id, r.content, r.is_positive;";
+
+        return jdbcTemplate.queryForObject(sqlQuery, this::makeReview, id);
+    }
+
+    public void updateReviewById(Review review) {
+        String sqlQuery = "UPDATE reviews SET content = ?, is_positive = ? WHERE review_id = ?";
+
+        jdbcTemplate.update(connection -> {
```

Comment:
А почему тут просто не использовать обычную форму (sql, param1, ...)? Какой-то вид оптимизации?

Reply:
Угу  
https://en.wikipedia.org/wiki/Prepared_statement 

> In database management systems (DBMS), a prepared statement, parameterized statement, or parameterized query is a feature used to pre-compile SQL code, separating it from data. Benefits of prepared statements are:
> 
> - efficiency, because they can be used repeatedly without re-compiling
> - security, by reducing or eliminating SQL injection attacks